# Regression:
* Linear Regression emphasis relation betweeen 2 variables and tries to fit relation through a line. Y = mX + C, X is the independent variable and Y is the dependent variable and C is the bias normalizer.
* Scatter Plot and Corelation coefficients are the popular metrics to determine relation beween 2 variables/features.
* Scatter plot would have x axis as independent and y axis as the one where would like to see the pattern of the target/dependent variable. It also indicates the direction of dependency.
* Corelation coefficients are the values between -1 to 1. where negative indicates indirect co relation and values closer to -1 and 1 indicates the max strength of the co relation.
* Least squarred error(popular method) is the way through which a best fit line is chosen.

In [1]:
import pandas as pd
import statsmodels.api as sm

In [2]:
df = pd.read_csv("train.csv")
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea    ...    MoSold YrSold SaleType SaleCondition SalePrice
0   1          60       RL         65.0     8450    ...         2   2008       WD        Normal    208500
1   2          20       RL         80.0     9600    ...         5   2007       WD        Normal    181500
2   3          60       RL         68.0    11250    ...         9   2008       WD        Normal    223500
3   4          70       RL         60.0     9550    ...         2   2006       WD       Abnorml    140000
4   5          60       RL         84.0    14260    ...        12   2008       WD        Normal    250000

[5 rows x 81 columns]

In [3]:
df[["GrLivArea","SalePrice"]].head()

GrLivArea  SalePrice
0       1710     208500
1       1262     181500
2       1786     223500
3       1717     140000
4       2198     250000

In [6]:
#adding intercept
df['intercept'] = 1

lm = sm.OLS(df['SalePrice'],df[['intercept','GrLivArea']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.502
Model:                            OLS   Adj. R-squared:                  0.502
Method:                 Least Squares   F-statistic:                     1471.
Date:                Mon, 25 Mar 2019   Prob (F-statistic):          4.52e-223
Time:                        19:43:29   Log-Likelihood:                -18035.
No. Observations:                1460   AIC:                         3.607e+04
Df Residuals:                    1458   BIC:                         3.608e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept   1.857e+04   4480.755      4.144      0.000    9779.612    2.74e+04
GrLivArea    107.1304      2.794     38.348      0.000     101.650     112.610
==============================================================================
Omnibus:                      261.166   Durbin-Watson:                   2.025
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3432.287
Skew:                           0.410   Prob(JB):                         0.00
Kurtosis:                      10.467   Cond. No.                     4.90e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.9e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Coef**
* intecept values indicate that when area is 0 the price of the house is 18570
* for one unit increase in area, sale price would increase by 107 dollar price.

**P>[t]**
* 0.000 (low value) indicates that the variable is important for SalePrice prediction

**R2 Squred**
* co relation between the variables (SalePrice and Area)

# Multiple Linear Regression

In [22]:
import numpy as np
df.fillna(method='ffill', inplace=True)
df_1 = df.drop('SalePrice',axis=1)
lm_2 = sm.OLS(df['SalePrice'], df_1[df_1.select_dtypes(include=np.number).columns])

In [23]:
result_2 = lm_2.fit()
result_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.813
Model:                            OLS   Adj. R-squared:                  0.809
Method:                 Least Squares   F-statistic:                     177.2
Date:                Mon, 25 Mar 2019   Prob (F-statistic):               0.00
Time:                        21:19:31   Log-Likelihood:                -17319.
No. Observations:                1460   AIC:                         3.471e+04
Df Residuals:                    1424   BIC:                         3.490e+04
Df Model:                          35                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Id               -1.1630      2.187     -0.532      0.595      -5.454       3.128
MSSubClass     -185.2600     27.532     -6.729      0.000    -239.268    -131.252
LotFrontage     -77.5950     45.929     -1.689      0.091    -167.692      12.502
LotArea           0.4332      0.102      4.258      0.000       0.234       0.633
OverallQual    1.735e+04   1188.185     14.603      0.000     1.5e+04    1.97e+04
OverallCond    4655.3876   1034.811      4.499      0.000    2625.470    6685.305
YearBuilt       281.0024     66.293      4.239      0.000     150.959     411.046
YearRemodAdd    136.5182     68.805      1.984      0.047       1.547     271.489
MasVnrArea       31.1692      5.947      5.241      0.000      19.503      42.836
BsmtFinSF1        9.9700      2.529      3.942      0.000       5.008      14.932
BsmtFinSF2       -0.9849      4.507     -0.219      0.827      -9.825       7.855
BsmtUnfSF         0.0413      2.411      0.017      0.986      -4.689       4.772
TotalBsmtSF       9.0264      3.366      2.681      0.007       2.423      15.630
1stFlrSF         18.4135      6.226      2.958      0.003       6.200      30.627
2ndFlrSF         18.0480      5.729      3.150      0.002       6.810      29.285
LowQualFinSF     -5.4224     15.007     -0.361      0.718     -34.861      24.016
GrLivArea        31.0391      5.706      5.439      0.000      19.845      42.233
BsmtFullBath   9297.5490   2611.137      3.561      0.000    4175.461    1.44e+04
BsmtHalfBath   1896.0671   4092.012      0.463      0.643   -6130.951    9923.085
FullBath       3531.5583   2831.061      1.247      0.212   -2021.939    9085.056
HalfBath      -1813.0252   2662.854     -0.681      0.496   -7036.562    3410.512
BedroomAbvGr  -9994.9937   1702.703     -5.870      0.000   -1.33e+04   -6654.918
KitchenAbvGr   -1.21e+04   5210.651     -2.322      0.020   -2.23e+04   -1875.630
TotRmsAbvGrd   5051.0126   1238.389      4.079      0.000    2621.750    7480.276
Fireplaces     3878.4959   1774.621      2.186      0.029     397.345    7359.647
GarageYrBlt     101.9752     64.196      1.589      0.112     -23.953     227.904
GarageCars     1.126e+04   2873.033      3.920      0.000    5626.388    1.69e+04
GarageArea       -3.5474      9.900     -0.358      0.720     -22.968      15.874
WoodDeckSF       23.5701      8.023      2.938      0.003       7.833      39.308
OpenPorchSF      -2.3247     15.151     -0.153      0.878     -32.046      27.397
EnclosedPorch    11.7238     16.859      0.695      0.487     -21.348      44.795
3SsnPorch        18.6589     31.398      0.594      0.552     -42.933      80.251
ScreenPorch      55.6276     17.189      3.236      0.001      21.910      89.346
PoolArea        -27.7994     23.774     -1.169      0.242     -74.434      18.836
MiscVal          -0.7854      1.855     -0.423   

* How do we get the coefficient of variables? 

\begin{equation*}
\beta =((X^TX)^{Inv})X^Ty
\end{equation*}
    * note that these are matrix multiplication
    
* linear equation is : y = b0 + b1X
* multilinear equation is : y = b0 + b1.X1 + b2.X2 + b3.X3